### PR TITLE
Feat/mappingderivs

### LINKED ISSUE #342: mapping.deriv
we should allow `maps.deriv` to take a vector and return a vector. Right now, we are passing around matrices (mostly sparse, so not really a big deal, but still it is unnecessary). If `v=None` we can simply return the matrix

### PR DESCRIPTION
**Pull request by lheagy**
mapping derivs can take a vector and return a vector or pass around matrices if `v=None`

This closes #342

### CODE REVIEW

**Comment by lheagy** [comment]
@rowanc1 : what do you think of this?

File: SimPEG/Maps.py [hunk 1]
```diff
@@ -90,13 +90,15 @@ def deriv(self, m):
...
+        if v is not None:
+            return v
         return sp.identity(self.nP)
 
     def test(self, m=None, **kwargs):
         """Test the derivative of the mapping.
 
             :param numpy.array m: model
-            :param kwargs: key word arguments of :meth:`SimPEG.Tests.checkDerivative`
+            :param kwargs: key word arguments of :math:`SimPEG.Tests.checkDerivative`
```
Comment (rowanc1) [inline SimPEG/Maps.py]: Is this math?

Comment (lheagy) [inline SimPEG/Maps.py]: ha, no it is not `meth` or `math`. This should be `ref` ?

**Comment by rowanc1** [comment]
This looks like a good addition!

**Comment by lheagy** [comment]
- [ ] `ParametricSplineMap` and 
- [ ] `ParametricPolyMap` 

are not currently tested. 
@sgkang : could you add tests for these?

**Comment by lheagy** [comment]
@sgkang : could you add tests for 
- [ ] parametric spline and
- [ ] parametric poly 
maps?

Also: is the 'logSigma' flag necessary? could this be accomplished with a combo map?

**Pull request merged**